Ar/algorithmic strategies (#34)
Ada-Developers-Academy/core-unit-3: algorithmic-strategies/dynamic-programming.md

| Dynamic Programming | An algorithmic strategy of breaking a problem down into subproblems that need to be calculated multiple times, allowing us to improve performance by storing results and reusing them. | "This algorithm calculates the same subproblem over and over, so let's use a dynamic-programming approach to improve the performance!" |
| Memoization         | An optimization technique used primarily to speed up algorithms by storing the results of subproblems and returning the cached result when the same subproblem occurs again.           | "If we memoize the results of that function call, we can improve the performance if it gets called again with the same arguments."     |

## Example: Fibonacci

The classic Fibonacci sequence is inherently recursive, but also inefficient to solve using the typical recursive manner.

In the Fibonacci sequence `fibonacci(n)`:

`fibonacci(0)` = 0, for \(n = 0\)  
`fibonacci(1)` = 1, for \(n = 1\)  
`fibonacci(n)` = `fibonacci(n-1)` + `fibonacci(n-2)`, for \(n > 1\)

> We could code the Fibonacci sequence as follows:

Author: Fixed a spacing issue in the suggestion.

```

However this is wildly inefficient. Note below for the Fibonacci of 5, how `fibonacci(2)` is called 3 times, and `fibonacci(1)` is called 5 times. As \(n\) grows larger, this will occur more and more often. For any \(n > 1\), we end up making \(2^n\) method calls!

![fibonacci(5) calls fibonacci(4)(a) and fibonacci(3)(b). fibonacci(4)(a) calls fibonacci(3)(c) and fibonacci(2)(d). fibonacci(3)(c) calls fibonacci(2)(e) and fibonacci(1)(f). fibonacci(2)(e) calls fibonacci(1)(g) and fibonacci(0)(h). fibonacci(2)(d) calls fibonacci(1)(i) and fibonacci(0)(j). fibonacci(3)(b) calls fibonacci(2)(k) and fibonacci(1)(l). fibonacci(2)(k) calls fibonacci(1)(m) and fibonacci(0)(n).](../assets/algorithmic-strategies_dynamic-programming_fibonacci-inefficient.png)  
_Fig. Fibonacci of 5. Notice how the same subproblems are repeatedly called!_

Instead of solving the same problems over and over again we can solve these problems by storing them in a _memo_ and using the stored subproblems to make calculating the larger problem more efficient.

Let's take a look at a few variations of `fibonacci` which make use of memoization to improve their performance.

> <br />

Author: Updated since not all of the memoizing objects are arrays (the version of the initial recursive implementation uses a dict to avoid needing to do any fancy array pre-allocation). Also added another sentence about why we're storing the results.

<details>

> <summary>Forward iterative solution. Starts from the base cases of [0, 1] building up to the desired number.</summary>

Author: By "forward," I meant starting from 0 and calculating up to the final solution, as opposed to the initial recursive solution, which starts at the end, and works back to bottom of the recursion. I didn't really need to add the term to the iterative solution, other than that it parallels the first of the dynamic programming recursive solutions. So really it just turn into a handy label that I could use to refer to these a little later on.

Open to other suggestions. Maybe I could change the order to more like:

```suggestion
<summary>Iterative solution that starts from the base cases of [0, 1] building up to the desired number. Because it starts from the lower Fibonacci numbers and counts up, we can refer to this as a forward iterative solution.</summary>
```
or maybe
```suggestion
<summary>Forward iterative solution. We call it a forward solution since it starts from the base cases of [0, 1] and then builds up to the desired result.</summary>
```

> <summary>Forward iterative solution. Starts from the base cases of [0, 1] building up to the desired number.</summary>

Author: But maybe for this one, simplest is best?

```suggestion
<summary>Iterative solution. Starts from the base cases of [0, 1] building up to the desired number.</summary>
```

```python
def fibonacci(n):
    if n == 0 or n == 1:
        return n

    solutions = [0, 1]
    current = 2

    while current <= n:
        solutions.append(solutions[current - 1] + solutions[current - 2])
        current += 1

    return solutions[n]
```

</details>

<details>

> <summary>Forward recursive solution. A recursive adaptation of the forward iterative solution.</summary>

Author: If we can solve the first comment, perhaps this will fall in line?

> <summary>Forward recursive solution. A recursive adaptation of the forward iterative solution.</summary>

Author: Maybe just simplify this one too, and then we just need a name for the other recursive version...
```suggestion
<summary>Recursive solution. A recursive adaptation of the iterative solution.</summary>
```

```python
def fibonacci_recursive(n, solutions=None, current=None):
    if solutions is None:
        solutions = [0, 1]
        current = 2

    solutions.append(solutions[current - 1] + solutions[current - 2])

    if n <= current:
        return solutions[n]

    return fibonacci_recursive(n, solutions, current + 1)
```

</details>

<details>

> <summary>Traditional recursive solution. Based on the typical branching recursive solution, but memoizes calculations as they are performed.</summary>

Author: Technically, both of the recursive solutions are memoizing. They just do the calculations in different directions (or rather, they run in the same direction, but they way they get queued up is different). The whole forward/traditional thing came from me wanting to have a way to quickly distinguish between these two recursive implementations (the original material didn't show this version, but due to how closely it mirrors the initial recursive implementation, I really wanted to include it!)

How about this?

```suggestion
<summary>"Fixed" recursive solution. Based on the initial inefficient recursive solution, but memoizes calculations as they are performed. Notice how similar this implementation is to that initial version. The addition of memoization preserves the core of the original logic, but improves the performance significantly!</summary>
```

```python
def fibonacci_recursive(n, solutions=None):
    if solutions is None:
        solutions = {}

    if n in solutions:
        return solutions[n]

    if n == 0 or n == 1:
        solutions[n] = n
    else:
        solutions[n] = (fibonacci_recursive(n - 1, solutions) +
            fibonacci_recursive(n - 2, solutions))

    return solutions[n]
```

</details>

To really appreciate the improvements that dynamic programming can bring, we can try running `fibonacci(40)` with the original, recursive implementation and note how long this takes to run. If we then run _any_ of the versions that use dynamic-programming techniques, we should see a significant improvement!

### !callout-info

## Fibonacci Challenges

> Notice how when solving `fibonacci(n)` we only need `fibonacci(n-1)` and `fibonacci(n-2)`. Using this observation, can we reduce our space complexity from \(O(n)\) to \(O(1)\) for the forward iterative solution? Would a similar approach improve the forward recursive solution?

Author: took out the forward references and rephrased the suggestion a bit to still try to be specific about to which recursive solution we are referring.

> Notice how when solving `fibonacci(n)` we only need `fibonacci(n-1)` and `fibonacci(n-2)`. Using this observation, can we reduce our space complexity from \(O(n)\) to \(O(1)\) for the forward iterative solution? Would a similar approach improve the forward recursive solution?

Author: merging these two suggestions, since GH won't let me batch apply them at the same time
```suggestion
Notice how when solving `fibonacci(n)`, the only solutions we need are `fibonacci(n-1)` and `fibonacci(n-2)`. Using this observation, can we reduce our space complexity from \(O(n)\) to \(O(1)\) for the iterative solution? Are we able to apply this strategy to the recursive version of that solution?
```

"So you didn't need to recount because you remembered there were eight!
  Dynamic programming is just a fancy way to say:
  'remembering stuff to save time later'"
```

## Example: Longest Common Subsequence

The longest common subsequence problem takes two strings `str1` and `str2`, and returns the length of their longest common subsequence.

A subsequence of a string is a new string generated from the original string with some characters (can be none) deleted without changing the relative order of the remaining characters. (eg, "ace" is a subsequence of "abcde" while "aec" is not). A common subsequence of two strings is a subsequence that is common to both strings.

If there is no common subsequence, return 0.

| `str1`    | `str2`      | Subsequence | Length (result) |
| --------- | ----------- | ----------- | --------------- |
| `"abcde"` | `"ace"`     | `"ace"`     | 3               |
| `"abcde"` | `"aqzcrwe"` | `"ace"`     | 3               |
| `"abc"`   | `"abc"`     | `"abc"`     | 3               |
| `"abc"`   | `"def"`     | `""`        | 0               |
| `"abc"`   | `""`        | `""`        | 0               |

### !callout-info

## From Genomics to Git Diffs

> Variations of the longest common subsequence problem power solutions for problems ranging from gene sequence comparisons to Git diffs!

Author: I got 99 problems, and lcs ain't one!

And how do we find the maximum subsequence of the remaining strings? Why with recursion!

![The call tree for finding the longest common subsequence for the strings "ace" and "ae". Because there are three branches at each level, it spreads very quickly. There are large repeated sections of the call tree.](../assets/algorithmic-strategies_dynamic-programming_lcs.png)  
_Fig. If the recursive explosion of Fibonacci seemed bad, get a load of this! Crossed-out nodes indicate calls that result in a 0 length due to at least one of the input strings being empty. The numbers indicate the maximum value being returned back from a particular branch of execution. Repeated regions are outlined in yellow._

> The explosion of calls in this diagram puts Fibonacci to shame! But for small examples, like in the example table above, it can be manageable with a traditional recursive implementation.

Author: reworded to avoid traditional
```suggestion
The explosion of calls in this diagram puts Fibonacci to shame! But for small examples, like in the example table above, it can be manageable with a recursive implementation like the following:
```

```python
def lcs(str1, str2):
    if not str1 or not str2:
        return 0

    # split the first character from the rest
    first1 = str1[0]
    rest1 = str1[1:]
    first2 = str2[0]
    rest2 = str2[1:]

    # is this spot a match?
    if first1 == first2:
        current_score = 1
    else:
        current_score = 0

    # the result for this position is the max of the current score
    # and each of the maxes from the three possibilities:
    # 1. advance both characters (adding to score if a match)
    # 2. advance only the first character
    # 3. advance only the second character
    result = max(
        current_score + lcs(rest1, rest2),
        lcs(rest1, str2),
        lcs(str1, rest2)
    )

    return result
```

> But increasing the input size can slow things down quickly! Consider calculating the longest common subsequence for `"tagacgttagtc"` and `"qaqaqgqtqgqc"`. The traditional recursive solution will take a noticeable amount of time.

Author: Here traditional is referring to it being un-memoized. I was trying to come up with a label other than simple (again simple as opposed to memoized), but simple could leave the reader feeling bad if it doesn't appear simple to them. So I'm looking for a descriptor that distinguishes this as "just the way we might attempt a recursive implementation in the absence of knowledge about memoization." So like calling it an un-memoized version seemed a bit strange, since why would we call this un-memoized if we didn't know that memoization was a thing?

> And how do we find the maximum subsequence of the remaining strings? Why with recursion!
> 
> ![The call tree for finding the longest common subsequence for the strings "ace" and "ae". Because there are three branches at each level, it spreads very quickly. There are large repeated sections of the call tree.](../assets/algorithmic-strategies_dynamic-programming_lcs.png)  
> _Fig. If the recursive explosion of Fibonacci seemed bad, get a load of this! Crossed-out nodes indicate calls that result in a 0 length due to at least one of the input strings being empty. The numbers indicate the maximum value being returned back from a particular branch of execution. Repeated regions are outlined in yellow._
> 
> The explosion of calls in this diagram puts Fibonacci to shame! But for small examples, like in the example table above, it can be manageable with a traditional recursive implementation.
> 
> ```python
> def lcs(str1, str2):
>     if not str1 or not str2:
>         return 0
> 
>     # split the first character from the rest
>     first1 = str1[0]
>     rest1 = str1[1:]
>     first2 = str2[0]
>     rest2 = str2[1:]
> 
>     # is this spot a match?
>     if first1 == first2:
>         current_score = 1
>     else:
>         current_score = 0
> 
>     # the result for this position is the max of the current score
>     # and each of the maxes from the three possibilities:
>     # 1. advance both characters (adding to score if a match)
>     # 2. advance only the first character
>     # 3. advance only the second character
>     result = max(
>         current_score + lcs(rest1, rest2),
>         lcs(rest1, str2),
>         lcs(str1, rest2)
>     )
> 
>     return result
> ```
> 
> But increasing the input size can slow things down quickly! Consider calculating the longest common subsequence for `"tagacgttagtc"` and `"qaqaqgqtqgqc"`. The traditional recursive solution will take a noticeable amount of time.
> 
> From the example call diagram, if we notice that there are many repeated sub-trees (especially the calls to `lcs("ce", "e")`, outlined in yellow), we might think about applying dynamic programming practices to reuse these calculations.

Reviewer: I think the flow might be interesting if we showed the recursive solution in code first, then showed the visualization of the spaghetti diagram, to drive the point that this is inefficient.

Also, the comment "The explosion of calls in this diagram puts Fibonacci to shame!" is a great one, but right before it, I'd love a more concrete observation acknowledged: there are a lot of function calls. (aka before calling it "explosive" (rightly so), name why it's explosive first)

What I did in this suggestion is add `<beginning of the code block haha>` and `<end code block haha>` around the recursive Python solution.

```suggestion
And how do we find the maximum subsequence of the remaining strings? Why with recursion!

<beginning of the code block haha>
def lcs(str1, str2):
    if not str1 or not str2:
        return 0

    # split the first character from the rest
    first1 = str1[0]
    rest1 = str1[1:]
    first2 = str2[0]
    rest2 = str2[1:]

    # is this spot a match?
    if first1 == first2:
        current_score = 1
    else:
        current_score = 0

    # the result for this position is the max of the current score
    # and each of the maxes from the three possibilities:
    # 1. advance both characters (adding to score if a match)
    # 2. advance only the first character
    # 3. advance only the second character
    result = max(
        current_score + lcs(rest1, rest2),
        lcs(rest1, str2),
        lcs(str1, rest2)
    )

    return result
<end code block haha>

However, let's visualize this solution in a diagram. In this example, we're finding the longest common subsequence for the strings "ace" and "ae." When we illustrate the number of function calls needed, we should see there are _a lot_.

![The call tree for finding the longest common subsequence for the strings "ace" and "ae". Because there are three branches at each level, it spreads very quickly. There are large repeated sections of the call tree.](../assets/algorithmic-strategies_dynamic-programming_lcs.png)  
_Fig. If the recursive explosion of Fibonacci seemed bad, get a load of this! Crossed-out nodes indicate calls that result in a 0 length due to at least one of the input strings being empty. The numbers indicate the maximum value being returned back from a particular branch of execution. Repeated regions are outlined in yellow._

The explosion of calls in this diagram puts Fibonacci to shame! But for small examples, like in the example table above, it can be manageable with a traditional recursive implementation.

But increasing the input size can slow things down quickly! Consider calculating the longest common subsequence for `"tagacgttagtc"` and `"qaqaqgqtqgqc"`. The traditional recursive solution will take a noticeable amount of time.

From the example call diagram, if we notice that there are many repeated sub-trees (especially the calls to `lcs("ce", "e")`, outlined in yellow), we might think about applying dynamic programming practices to reuse these calculations.
```

Reviewer: For convenience, here is the suggestion above, with the code snippet ripped out:

---

And how do we find the maximum subsequence of the remaining strings? Why with recursion!

However, let's visualize this solution in a diagram. In this example, we're finding the longest common subsequence for the strings "ace" and "ae." When we illustrate the number of function calls needed, we should see there are _a lot_.

![The call tree for finding the longest common subsequence for the strings "ace" and "ae". Because there are three branches at each level, it spreads very quickly. There are large repeated sections of the call tree.](../assets/algorithmic-strategies_dynamic-programming_lcs.png)  
_Fig. If the recursive explosion of Fibonacci seemed bad, get a load of this! Crossed-out nodes indicate calls that result in a 0 length due to at least one of the input strings being empty. The numbers indicate the maximum value being returned back from a particular branch of execution. Repeated regions are outlined in yellow._

The explosion of calls in this diagram puts Fibonacci to shame! But for small examples, like in the example table above, it can be manageable with a traditional recursive implementation.

But increasing the input size can slow things down quickly! Consider calculating the longest common subsequence for `"tagacgttagtc"` and `"qaqaqgqtqgqc"`. The traditional recursive solution will take a noticeable amount of time.

From the example call diagram, if we notice that there are many repeated sub-trees (especially the calls to `lcs("ce", "e")`, outlined in yellow), we might think about applying dynamic programming practices to reuse these calculations.

To add memoization to the Fibonacci algorithm, we only had to check whether a single value had been calculated before. But for the longest common subsequence we need to look up a result based on two inputs.

<br />

<details>

> <summary>Think about how we might accomplish this, and then click here to review a possible solution. Comments are included to indicate the newly added code.</summary>

Reviewer: For situations like this, I randomly prefer "Consider" over "Think about."

For me, "thinking about" feels more like musing, like it should be casual, while "considering" has that _weight_, but in a casual way.

Split the first sentence to give the "Consider" step a nice pause.

I won't fight for this suggestion :)

```suggestion
<summary>Consider how we might accomplish this. Then click here to review a possible solution. Comments are included to indicate the newly added code.</summary>
```
